ceph: perfcounter priorities and daemonperf updates to use them

dmick opened the pull request:
Builds on Sage's perfcounter priorities branch, adds to daemonperf.  

New features:

- [stat-pats]:   optional list of shell-glob patterns to select stats
- [priority]:   optional minimum priority of stat to allow (default: 'useful')
- daemonperf list|ls: Show tabular list of selected stats and their nicknames
- daemonperf display now reacts when terminal is resized, truncating sections if necessary

liewegas commented on src/ceph.in:
```diff
@@ -664,6 +668,10 @@ def daemonperf(childargs, sockpath):
     while len(childargs) and not isnum(childargs[0]):
         arg = childargs.pop(0)
+        # 'list'?
+        if arg == 'list':
```
can we make 'ls' a synonym, since that's what we tend to use elsewhere?

dmick replied:
sure

liewegas commented:
<pre>
gnit:build (master) 10:03 AM $ bin/ceph -s
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
Traceback (most recent call last):
  File "bin/ceph", line 126, in <module>
    from ceph_daemon import DaemonWatcher, admin_socket
  File "/home/sage/src/ceph3/src/pybind/ceph_daemon.py", line 21, in <module>
    from prettytable import PrettyTable, HEADER
ImportError: No module named prettytable
</pre>
i re-ran sudo ./install-deps.sh so this is presuambly a missing package dependency?

liewegas commented:
Hmm, we could also make the daemonperf ls skip things with prio=0 (and/or no nicks)

dmick commented:
Hm.  It was already used in mgr/fsstatus, but I guess it hasn't been added to dependencies yet.  Will do or find a substitute.

dmick commented:
Incorporated most of https://github.com/dmick/ceph/pull/1 with a fix or two, added 'ls' as suggested; still want to verify prettytable is universally available and add to package deps

dmick commented:
also, clearly, I must fix the test code I hadn't realized existed

dmick commented:
OK, tests fixed, build and runtime dependencies added

tchaikov commented on ceph.spec.in:
```diff
@@ -121,6 +121,7 @@ BuildRequires:	pkgconfig
 BuildRequires:	python
 BuildRequires:	python-devel
 BuildRequires:	python-nose
+BuildRequires:	python-prettytable
```
do we run `ceph` at build-time? "make_check" is disabled by default. maybe we can guard this line like
```
%if 0%{with make_check}
BuildRequires:	python-prettytable
%endif
```
?

dmick replied:
Well, I assume the same thing applies as for Debian, in that the package needs to be installed for make check to work.  Since the make check builds don't install the built packages to pull in prettytable via Requires, there's nothing else to do it that I know of.  (if there is some sort of "make check dependencies" mechanism I'm happy to learn about it.)

tchaikov replied:
so the consumer of this line is install-deps.sh actually. i am far from an expert of `dnf` or its builddep plugin. so guess this is the best way to document the "make check" dependencies at this moment.

tchaikov commented on debian/control:
```diff
@@ -50,6 +50,7 @@ Build-Depends: bc,
                python (>= 2.7),
                python-all-dev,
                python-nose,
+	       python-prettytable,
```
debian packaging does not run "make check", so we can remove `python-prettytable` from Build-Depends.

dmick replied:
The thing is, if it's not in build-depends, make check fails, because the target host doesn't have the package to run the tests.

liewegas replied:
Yeah, I think it needs to be here.  The question I have is whether the debbuilder stuff magically detects this python dependency as a package dependency too (the way it automatically figures out shared library dependencies) or whether the python dependency has to be explicitly declared...

tchaikov replied:
agreed. we debated about "make check" dependencies machinery before, iirc. but no acceptable result back then. i should learn to live with it.

tchaikov commented on src/ceph.in:
```diff
@@ -266,10 +275,15 @@ ping <mon.id>           Send simple presence/life test to a mon
+                        List shows a table of all available stats
                         Run <count> times (default forever),
-                        once per <interval> seconds (default 1)
+                         once per <interval> seconds (default 1)
```
why added an extra space here?

dmick replied:
to show that it's part of the preceding line more clearly

tchaikov commented on src/ceph.in:
```diff
@@ -565,6 +579,133 @@ def ping_monitor(cluster_handle, name, timeout):
+    if len(childargs) > 0:
+        try:
+            interval = float(childargs.pop(0))
+            if interval < 0:
```
nit, why not just print the error message and return `errno.EINVAL`?

dmick replied:
because the exception can also come from float() (bad format), and this way both errors exit the same way

tchaikov replied:
oic.

liewegas commented:
In a narrow window for daemonperf osd.0, it shows all of the bluefs metrics (including the low priority ones) but not the higher-priority osd ones.

dmick commented:
Not sure what you mean.   Default priority is 'useful', 5, and the ones it shows are all the ones with 5 or greater:


```
| section   | name               | nick | prio |
+-----------+--------------------+------+------+
| bluefs    | db_total_bytes     | b    |    5 |
| bluefs    | db_used_bytes      | u    |    5 |
| bluefs    | wal_total_bytes    | walb |    5 |
| bluefs    | wal_used_bytes     | walu |    5 |
| bluefs    | slow_total_bytes   | slob |    5 |
| bluefs    | slow_used_bytes    | slou |    5 |
| bluefs    | num_files          | f    |    5 |
| bluefs    | log_bytes          | jlen |    8 |
| bluefs    | logged_bytes       | j    |   10 |
| bluefs    | bytes_written_wal  | wal  |   10 |
| bluefs    | bytes_written_sst  | sst  |   10 |
| bluestore | kv_flush_lat       | fl_l |    8 |
| bluestore | kv_lat             | k_l  |    8 |
| bluestore | state_aio_wait_lat | io_l |    8 |
| bluestore | throttle_lat       | th_l |   10 |
| bluestore | submit_lat         | s_l  |   10 |
| bluestore | commit_lat         | c_l  |   10 |
| bluestore | read_lat           | r_l  |   10 |
| objecter  | op_active          | actv |    7 |
| objecter  | op_r               | rd   |    7 |
| objecter  | op_w               | wr   |    7 |
| objecter  | op_rmw             | rdwr |    5 |
| osd       | op                 | ops  |   10 |
| osd       | op_in_bytes        | wr   |    8 |
| osd       | op_out_bytes       | rd   |    8 |
| osd       | op_latency         | l    |    9 |
| osd       | recovery_ops       | rop  |    8 |
| osd       | numpg              | pgs  |    5 |

```

dmick commented:
Are you suggesting that you'd rather have it sort by priority?...because of the sections that would be hard

liewegas commented:
oh i see.  i was imagining that it could look at the terminal width and 
figure out the lowest priority level that fit everything in the terminal.

liewegas commented:
...but still leave the ordering the same.

dmick commented:
Ah.  that would be pretty slick and surprising IMO, but anything's possible of course.  You can, however, use patterns to select more specifically.

dmick commented:
retest this please

dmick commented:
@liewegas: anything else here?
